packages/puppeteer-core/src/util/Deferred.ts

```diff
@@ -17,26 +17,28 @@ export interface DeferredOptions {
  *
  * @internal
  */
-export class Deferred<T> {
+export class Deferred<T, V extends Error = Error> {
   #isResolved = false;
   #isRejected = false;
-  #value: T | Error | undefined;
+  #value: T | V | TimeoutError | undefined;
   #resolver: (value: void) => void = () => {};
   #taskPromise = new Promise<void>(resolve => {
     this.#resolver = resolve;
   });
   #timeoutId: ReturnType<typeof setTimeout> | undefined;
+  #timeoutError: TimeoutError;
 
   constructor(opts?: DeferredOptions) {
+    this.#timeoutError = new TimeoutError(opts && opts.message);
```

Comment:
Create the error on init so the StackTrace have the path to where it was called similar to Callbacks in protocol.